### PR TITLE
Fixed #26430, Refs #24462 -- Added BaseSubquery datastructure.

### PR DESCRIPTION
**Pull request by claudep**
Added new Join-like class BaseSubquery which is meant to be used in
FROM clauses. Rewrote Query.get_aggregation with it. Deprecated
AggregateQuery and SQLAggregateCompiler as redundant.

### CODE REVIEW

File: tests/aggregation_regress/tests.py [hunk 1]
```diff
@@ -972,10 +972,11 @@ def test_empty_filter_count(self):
         )
 
     def test_empty_filter_aggregate(self):
-        self.assertEqual(
-            Author.objects.filter(id__in=[]).annotate(Count("friends")).aggregate(Count("pk")),
-            {"pk__count": None}
-        )
+        with self.assertNumQueries(0):
```
Comment (charettes) [inline tests/aggregation_regress/tests.py]: Awesome!

File: django/db/models/sql/datastructures.py [hunk 2]
```diff
@@ -166,3 +166,25 @@ def equals(self, other, with_filtered_relation):
...
+
+
+class BaseSubquery:
+    join_type = None
+    parent_alias = None
+    filtered_relation = None
+
+    def __init__(self, query, alias):
+        self.query = query
+        self.query.subquery = True
```
Comment (charettes) [inline django/db/models/sql/datastructures.py]: That's the only thing that I think we should change here is too make the caller assign `subquery = True` and assert `query.subquery` here. Altering an argument is usually bad practice so I think `get_aggregation` should assign `inner_query.subquery = True` instead.

**Comment by claudep** [comment]
Simon, I think it would be really better if you could take ownership of the patch. I would really like to help, but doing thing blindly for which I don't understand implications is not wise.

**Comment by claudep** [comment]
@gizmondo, do you have any following patches, even POCs, to reach #24462 goals?

**Comment by gizmondo** [comment]
@claudep, first of all, thanks for trying to revive this effort! 

There were some experimental patches, but unfortunately their history is in a private repo to which I don't have access anymore. Nothing particularly valuable there anyway, afair.

File: django/db/models/sql/datastructures.py [hunk 3]
```diff
@@ -166,3 +166,25 @@ def equals(self, other, with_filtered_relation):
...
+
+
+class BaseSubquery:
+    join_type = None
+    parent_alias = None
+    filtered_relation = None
+
+    def __init__(self, query, alias):
+        self.query = query
+        assert self.query.subquery is True
```
Comment (atombrella) [inline django/db/models/sql/datastructures.py:178]: This is probably still a bit WIP, but can we try to use `raise ValueError(...)` instead?

**Comment by claudep** [comment]
It think this PR is partly obsolete. @charettes, if you think anything here is worth resurrecting, please speak!

**Pull request closed without merging**